crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+    { OSSL_HPKE_KDF_ID_HKDF_SHA384, LN_sha384, SHA384_DIGEST_LENGTH },
+    { OSSL_HPKE_KDF_ID_HKDF_SHA512, LN_sha512, SHA512_DIGEST_LENGTH }
+};
+
+/**
+ * @brief sender or receiver context
+ */
+struct ossl_hpke_ctx_st
+{
+    OSSL_LIB_CTX *libctx; /**< library context */
+    char *propq; /**< properties */
+    int mode; /**< HPKE mode */
+    OSSL_HPKE_SUITE suite; /**< suite */
+    uint64_t seq; /**< sequence number */
+    unsigned char *shared_secret;
+    size_t shared_secretlen;
+    unsigned char *exportersec; /**< exporter secret */
+    size_t exporterseclen;
+    char *pskid; /**< PSK stuff */
+    unsigned char *psk;
+    size_t psklen;
+    EVP_PKEY *senderpriv; /**< sender's ephemeral private key */
+    char *ikme;
+    size_t ikmelen;
+    EVP_PKEY *authpriv; /**< sender's authentication private key */
+    unsigned char *authpub; /**< auth public key */
+    size_t authpublen;
+};
+
+/*
+ * @brief map from IANA codepoint to AEAD table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in AEAD table or 0 if error
+ */
+static uint16_t aead_iana2index(uint16_t codepoint)
+{
+    uint16_t naeads = OSSL_NELEM(hpke_aead_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != naeads; i++) {
+        if (hpke_aead_tab[i].aead_id == codepoint)
+            return i;
+    }
+    return 0;
+}
+
+/*
+ * @brief map from IANA codepoint to KEM table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in KEM table or 0 if error
+ */
+static uint16_t kem_iana2index(uint16_t codepoint)
+{
+    uint16_t nkems = OSSL_NELEM(hpke_kem_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != nkems; i++) {
+        if (hpke_kem_tab[i].kem_id == codepoint)
+            return i;
+    }
+    return 0;
+}
+
+/*
+ * @brief map from IANA codepoint to AEAD table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in AEAD table or 0 if error
+ */
+static uint16_t kdf_iana2index(uint16_t codepoint)
+{
+    uint16_t nkdfs = OSSL_NELEM(hpke_kdf_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != nkdfs; i++) {
+        if (hpke_kdf_tab[i].kdf_id == codepoint)
+            return i;
+    }
+    return 0;
+}
+
+/*
+ * @brief Check if kem_id is ok/known to us
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for good, not 1 for error
+ */
+static int hpke_kem_id_check(uint16_t kem_id)
+{
+    switch (kem_id) {
+    case OSSL_HPKE_KEM_ID_P256:
+    case OSSL_HPKE_KEM_ID_P384:
+    case OSSL_HPKE_KEM_ID_P521:
+    case OSSL_HPKE_KEM_ID_X25519:
+    case OSSL_HPKE_KEM_ID_X448:
+        break;
+    default:
+        return 0;
+    }
+    return 1;
+}
+
+/*
+ * @brief check if KEM uses NIST curve or not
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for NIST, 0 for good-but-non-NIST, other otherwise
+ */
+static int hpke_kem_id_nist_curve(uint16_t kem_id)
+{
+    switch (kem_id) {
+    case OSSL_HPKE_KEM_ID_P256:
```

Comment:
Similiar comment to above related to table search

Reply:
will close this so - enough to keep one open for discussion